### ISSUE TITLE
atom map for some reaction  is error, how to fix this probrom

### ISSUE THREAD

**Issue by xiaoqiangsheng2016** [rxn4chemistry/rxnmapper#12]
for this reaction
 `OB(C1=CCCCC1)O.ClC2=CC=CC=C2>>C3(C4=CC=CC=C4)=CCCCC3`

the output mapping is:
` Cl[c:3]1[cH:4][cH:5][cH:6][cH:7][cH:8]1.OB(O)[C:1]1=[CH:2][CH2:9][CH2:10][CH2:11][CH2:12]1>>[CH:1]1=[C:2]([c:3]2[cH:4][cH:5][cH:6][cH:7][cH:8]2)[CH2:9][CH2:10][CH2:11][CH2:12]1`

![20210610215727](https://user-images.githubusercontent.com/24220486/121537942-f73dcf80-ca36-11eb-9102-358cc47697ef.png)

**Comment by pschwllr** [comment]
You're right -  the mapping is wrong in this specific case. The beauty of the approach is that it learned to map reactions without unsupervised (without having labelled examples). That said, it is not straightforward to improve and correct the model. Reporting wrongly mapped reactions like you just did is very helpful, as we will consider them for the next RXNMapper version. 

Let me show explain to you what happens by looking at the attention weights on the [RXNMapper Demo](http://rxnmapper.ai/demo.html?rxn=OB(C1%253DCCCCC1)O.ClC2%253DCC%253DCC%253DC2%253E%253EC3(C4%253DCC%253DCC%253DC4)%253DCCCCC3&selectedLayer=10&selectedHead=5&selectedTokenSide=source&selectedTokenInd=29). If we look at the product atoms 1 and 2, and their attention weights towards the reactant atoms:
<img width="888" alt="image" src="https://user-images.githubusercontent.com/38880871/121665924-02314800-caa9-11eb-9b0b-7c625be445f8.png">
_Attention of product atom 1._

<img width="893" alt="image" src="https://user-images.githubusercontent.com/38880871/121665793-ddd56b80-caa8-11eb-9377-69d3f2d88e46.png">
_Attention of product atom 2._

We see that atom 2 has the stronger attention weight, and will be mapped first to the wrong reactant atom. 

One reason for this attention pattern is how the SMILES are canonicalised for this specific case. If we look at the wrongly mapped part of the SMILES, we have the following `... C 1 = C ... >> C 1 = C ...`.

This is incorrectly mapped by RXNMapper to  `... [C:1]1=[CH:2] ... >> [C:1]1=[CH:2] ...`. Due to the canonicalisation, in this case, the two carbon in the SMILES switched side around the `1 =`. 
The correct mapping would be `... [C:2]1=[CH:1] ... >> [C:1]1=[CH:2] ...`.

Although your example is straightforward to map for a human, it is challenging for an AI model that sees the molecules only as SMILES strings.

**Comment by feiranl** [comment]
Hi, I also want to report another case which is wrongly mapped.
rxn = `L-glutamine <=> NH4(+) + L-glutamate`
Smiles = `NC(=O)CC[C@H]([NH3+])C(=O)[O-]>>[NH4+].[NH3+][C@@H](CCC(=O)[O-])C(=O)[O-]`
mapped rxn = `[NH3+:1][C@@H:2]([CH2:3][CH2:4][C:5](=[O:6])[NH2:11])[C:8](=[O:7])[O-:10]>>[NH3+:1][C@@H:2]([CH2:3][CH2:4][C:5](=[O:6])[O-:7])[C:8](=[O:9])[O-:10].[NH4+:11]`
![image](https://user-images.githubusercontent.com/32157802/157904017-6025767a-c85a-4e3d-b3eb-8f8a727bb232.png)
The change of `O7` is not correct

**Comment by pschwllr** [comment]
With the default parameters, RXNMapper will map all product atoms iteratively (as we assume that all atoms are present on the precursors side). The problem is that the last product oxygen atom to map overwrites another previously mapped oxygen atom, as you omitted `H2O` in the reaction. If you add the `H2O`, I guess it maps it correctly:

<img width="745" alt="image" src="https://user-images.githubusercontent.com/38880871/163845246-c6b16b7f-1c1f-4056-92ad-d57b19530f6e.png">

**Comment by avaucher** [comment]
Closing due to inactivity.
